### ISSUE TITLE
Suggestion: add the option to ignore zeros in ReflectDataset?

### ISSUE THREAD

**Issue by James-Durant** [refnx/refnx#522]
Hi there,

When trying to fit data with points of zero reflectivity, understandably, NaNs are produced in the `Objective.logl` method and the following runtime error is raised:
![image](https://user-images.githubusercontent.com/43757679/102608868-79050e00-4122-11eb-9a4c-c4ce02bbf4d7.png)

Perhaps adding a method to the `ReflectDataset` class to ignore zeros or to remove them would be beneficial?

I'm currently using the following as a workaround:
```
dataset = np.loadtxt("test.dat")
dataset = dataset[(dataset != 0).all(1)]  #Remove any points with 0 reflectivity
q, r, r_error = dataset[:,0], dataset[:,1], dataset[:,2]
data = ReflectDataset([q, r, r_error])
```

Thanks,
James

**Comment by arm61** [comment]
Check out the `mask` property in the `Data1D` class. It might allow for a shorter workaround?

**Comment by James-Durant** [comment]
Ah yes, I forgot about that! Something like this works nicely:
```
data = ReflectDataset("test.dat") 
data.mask = (data.y != 0)
```

I raised the issue as, for someone new to refnx, it might not necessarily be obvious why they could be getting NaNs from fitting their data.

Anyway thank you for the suggestion Andrew,
James

**Comment by arm61** [comment]
Potentially worth making a PR where an error is raised if any of the values or uncertainties in a dataset are 0, with the recommendation to apply a `mask`?

**Comment by andyfaff** [comment]
James-Durant, thank you for contributing to refnx for the first time.

Experimental reflectivity curves may have points that have R=0, or even R < 0. With a perfect background subtraction one would expect half the points in that region to be above 0, and half below. Note - I would expect `sigma_R`, the uncertainty on the data points to be strictly greater than 0, `sigma_R > 0`.

If no `Transform` is used (fitting linY vs X), then this should be fine. However, there will be issues experienced if a logY vs X transform is used, as it can't cope with R <=0. 

People frequently use linY vs X, and it's therefore the wrong thing to do to filter those datapoints when they're loaded. It happens so frequently that it might be annoying to raise a warning when the dataset is loaded.

My preference would be for the `logY` transform (~L1227 in objective.py) to detect whether any of the transformed y are non-finite, and to raise the `RuntimeError` there. This error can be modified to something along the lines of "some of the transformed data was non-finite. Please check your datasets for points with zero or negative values". I think this is the right thing to do here.

Another alternate is to just do a `nansum` in logl, but this is covering over the original issue.

**Comment by arm61** [comment]
Oh that is a better idea than mine actually!

**Comment by James-Durant** [comment]
Ah, that sounds like a great idea. Having warnings raised all the time when they might not be relevant would be quite annoying.

**Comment by andyfaff** [comment]
James, would you like to make this addition to Transform, or would you prefer me to add?

**Comment by James-Durant** [comment]
Sure, I'll give it a go. It's all good experience

**Comment by andyfaff** [comment]
@James-Durant,  #524 raised a `RuntimeError` if any of the transformed points were non-finite. I've now changed that to a `RuntimeWarning` in #527. This is because the modification caused problems with the GUI operation, and to fix those would've involved jumping through a few hoops. Hopefully a warning will provide enough information. I've also reinstated a `RuntimeError` if the `Objective.logl` method experiences any non-finite values. Hopefully both of those will be indication enough that there are problems with the data.
